git-future-commit-workflow.md

```diff
@@ -0,0 +1,119 @@
+# Git "Future-Commit" Workflow
+
+## Motivation
+
+It's not uncommon for some feature to depend on the work of another feature on a separate branch. That feature's dependency may be in peer-review with an outstanding pull-request, or still a work-in-progress and not yet ready for review. Regardless, how can we build ontop of that feature's dependency in a clean way that is relatively simple to manage and clear for the reviewer to follow?
+
+This convention describes a workflow using "future commits" to organize and identify feature dependencies that have not yet been merged upstream (master branch). This is accomplished with a single commit for each anticipated merge of a feature, building a psuedo-master branch to be used until all feature dependencies (PRs) have been reviewed, approved, and merged.
+
+## Creating a "future merge commit"
+
+Say we have an existing feature branch, `feature-a`, that has yet to be merged into master, and we have another feature, `feature-b`, that we wish to have built on top of `feature-a`. We can do so by creating a psuedo-merge commit on master and build our `feature-b` from that commit.
+
+```sh
+git checkout master
+git checkout -b feature-b
+git merge --no-ff -m"Future merge of 'feature-a'" feature-a
+git tag -f future/feature-a
+```
+
+This will first checkout master, create our second feature branch (`feature-b`), merge our dependency feature branch (`feature-a`) into our new branch, and then tag the merge commit with a conventional `future/feature-a` tag for a convenient reference to this point in history.
+
+Now we're ready to make commits to `feature-b`! When we submit our PR, the reviewer can see our merge commit and know that all the work prior to this merge commit is to be reviewed separately with another PR.
+
+## Multiple future merges
+
+Let's say we're in the situation where we have a third feature which depends on `feature-b`. We can create a future commit for this using the same technique.
+
+```sh
+git checkout master
+git checkout -b feature-c
+git merge --no-ff -m"Future merge of 'feature-b'" feature-b
+git tag -f future/feature-b
+```
+
+This time we created our new branch `feature-c` and merged `feature-b`. Finally we tagged this new merge commit with `future/feature-b`. 
+
+Notice, we're always starting with master, creating a new branch, and merging the feature dependencies. We could turn this pattern into a git alias for convenience:
+
+```
+[alias]
+  timetravel = "!f(){\
+      git checkout master;\
+      git checkout -b $1;\
+      git merge --no-ff --no-verify -m"Future merge of '$2'" $2;\
+      git tag -f future/$2;\
+    }; f"
+```
+
+Add this to your `~/.gitignore` file. We can use it like so:
```

Comment:
You mean `~/.config/git/config` ?

Reply:
Good catch. I meant `~/.gitconfig` which I assume is the same thing.